Servers/DHCPClient/DHCPv4.h

```diff
@@ -0,0 +1,294 @@
+
+enum class DHCPv4Op : u8 {
+    BootRequest = 1,
+    BootReply = 2
+};
+
+enum class DHCPOption : u8 {
+    // BOOTP
+    Pad = 0,
+    SubnetMask,
+    TimeOffset,
+    Router,
+    TimeServer,
+    NameServer,
+    DomainNameServer,
+    LogServer,
+    CookieServer,
+    LPRServer,
+    ImpressServer,
+    ResourceLocationServer,
+    HostName,
+    BootFileSize,
+    MeritDumpFile,
+    DomainName,
+    SwapServer,
+    RootPath,
+    ExtensionsPath,
+    IPForwardingEnableDisable,
+    NonLocalSourceRoutingEnableDisable,
+    PolicyFilter,
+    MaximumDatagramReassemblySize,
+    DefaultIPTTL,
+    PathMTUAgingTimeout,
+    PathMTUPlateauTable,
+    InterfaceMTU,
+    AllSubnetsAreLocal,
+    BroadcastAddress,
+    PerformMaskDiscovery,
+    MaskSupplier,
+    PerformRouterDiscovery,
+    RouterSolicitationAddress,
+    StaticRoute,
+    TrailerEncapsulation,
+    ARPCacheTimeout,
+    EthernetEncapsulation,
+    TCPDefaultTTL,
+    TCPKeepaliveInterval,
+    TCPKeepaliveGarbage,
+    NetworkInformationServiceDomain,
+    NetworkInformationServers,
+    NetworkTimeProtocolServers,
+    VendorSpecificInformation,
+    NetBIOSOverTCPIPNameServer,
+    NetBIOSOverTCPIPDatagramDistributionServer,
+    NetBIOSOverTCPIPNodeType,
+    NetBIOSOverTCPIPScope,
+    XWindowSystemFontServer, // wow
+    XWindowSystemDisplayManager,
+    // DHCP
+    RequestedIPAddress = 50,
+    IPAddressLeaseTime,
+    OptionOverload,
+    DHCPMessageType,
+    ServerIdentifier,
+    ParameterRequestList,
+    Message,
+    MaximumDHCPMessageSize,
+    RenewalT1Time,
+    RenewalT2Time,
+    ClassIdentifier,
+    ClientIdentifier,
+    End = 255
+};
+
+enum class DHCPMessageType : u8 {
+    DHCPDiscover = 1,
+    DHCPOffer,
+    DHCPRequest,
+    DHCPDecline,
+    DHCPAck,
+    DHCPNak,
+    DHCPRelease,
+};
+
+template <>
+struct AK::Traits<DHCPOption> : public AK::GenericTraits<DHCPOption> {
+    static constexpr bool is_trivial() { return true; }
+    static unsigned hash(DHCPOption u) { return int_hash((u8)u); }
+};
+
+struct ParsedDHCPv4Options {
+    template <typename T>
+    Optional<const T> get(DHCPOption option_name) const
+    {
+        auto opt = options.get(option_name);
+        if (!opt.has_value()) {
+            return {};
+        }
+        auto& val = opt.value();
+        if (val.length != sizeof(T))
+            return {};
+        return *(const T*)val.value;
+    }
+
+    template <typename T>
+    Vector<T> get_many(DHCPOption option_name, size_t max_number) const
+    {
+        Vector<T> values;
+
+        auto opt = options.get(option_name);
+        if (!opt.has_value()) {
+            return {};
+        }
+        auto& val = opt.value();
+        if (val.length < sizeof(T))
+            return {};
+
+        for (size_t i = 0; i < max_number; ++i) {
+            auto offset = i * sizeof(T);
+            if (offset >= val.length)
+                break;
+            values.append(*(T*)((u8*)const_cast<void*>(val.value) + offset));
```

Comment:
This looks like it's missing bounds checking(?) Or maybe I'm not seeing it.

Reply:
It's not.
a little above, we assert that we have enough data for at least one element of size T, and here we simply go out of the loop if we run out of data.
`val.value` is guaranteed to be `val.length` bytes long

Reply:
Yeah no, awake me can't find a reason for this to access out of bounds either.